lms/services/vitalsource/service.py

```diff
@@ -65,22 +69,65 @@ def get_book_reader_url(cls, document_url) -> str:
 
         return f"https://hypothesis.vitalsource.com/books/{loc.book_id}/cfi/{loc.cfi}"
 
-    def get_sso_redirect(self, user_reference, document_url) -> str:
+    def get_sso_redirect(self, document_url, lti_params: LTIParams) -> str:
         """
         Get the public URL for VitalSource book viewer from our internal URL.
 
         That URL can be used to load VitalSource content in an iframe like we
         do with other types of content.
 
-        :param user_reference: The reference of the user
         :param document_url: `vitalsource://` type URL identifying the document
+        :param lti_params: Params to read the user reference from
         :raises VitalSourceError: If the user has no licences for the material
+            or if the service is misconfigured.
         """
         loc = VSBookLocation.from_document_url(document_url)
 
+        user_reference = self._get_user_reference(lti_params)
+
         if not self._sso_client.get_user_book_license(user_reference, loc.book_id):
             raise VitalSourceError("vitalsource_no_book_license")
 
         return self._sso_client.get_sso_redirect(
             user_reference, self.get_book_reader_url(document_url)
         )
+
+    @staticmethod
+    def compile_user_lti_pattern(pattern: str) -> Optional[re.Pattern]:
```

Comment:
I've made this public to make it easy to test without getting told off, but also because I thought it might be of interest to the admin pages if we end up checking things more thoroughly there.